x-pack/filebeat/input/entityanalytics/provider/okta/internal/okta/ratelimiter.go

```diff
@@ -16,33 +16,66 @@ import (
 	"github.com/elastic/elastic-agent-libs/logp"
 )
 
-type RateLimiter map[string]*rate.Limiter
+type RateLimiter struct {
+	window     time.Duration
+	fixedLimit *int
+	byEndpoint map[string]endpointRateLimiter
+}
+
+type endpointRateLimiter struct {
+	limiter *rate.Limiter
+	ready   chan struct{}
+}
 
-func NewRateLimiter() RateLimiter {
-	r := make(RateLimiter)
-	return r
+const waitDeadline = 30 * time.Minute
+
+func NewRateLimiter(window time.Duration, fixedLimit *int) *RateLimiter {
+	endpoints := make(map[string]endpointRateLimiter)
+	r := RateLimiter{
+		window:     window,
+		fixedLimit: fixedLimit,
+		byEndpoint: endpoints,
+	}
+	r.fixedLimit = fixedLimit
+	return &r
 }
 
-func (r RateLimiter) limiter(path string) *rate.Limiter {
-	if existing, ok := r[path]; ok {
+func (r RateLimiter) endpoint(path string) endpointRateLimiter {
+	if existing, ok := r.byEndpoint[path]; ok {
 		return existing
 	}
-	initial := rate.NewLimiter(1, 1) // Allow a single fetch operation to obtain limits from the API
-	r[path] = initial
-	return initial
+	limit := rate.Limit(1)
+	if r.fixedLimit != nil {
+		limit = rate.Limit(float64(*r.fixedLimit) / r.window.Seconds())
+	}
+	limiter := rate.NewLimiter(limit, 1) // Allow a single fetch operation to obtain limits from the API
+	ready := make(chan struct{})
+	close(ready)
+	newEndpointRateLimiter := endpointRateLimiter{
+		limiter: limiter,
+		ready:   ready,
+	}
+	r.byEndpoint[path] = newEndpointRateLimiter
+	return newEndpointRateLimiter
 }
 
 func (r RateLimiter) Wait(ctx context.Context, endpoint string, url *url.URL, log *logp.Logger) (err error) {
-	limiter := r.limiter(endpoint)
-	log.Debugw("rate limit", "limit", limiter.Limit(), "burst", limiter.Burst(), "url", url.String())
-	return limiter.Wait(ctx)
+	e := r.endpoint(endpoint)
+	<-e.ready
+	log.Debugw("rate limit", "limit", e.limiter.Limit(), "burst", e.limiter.Burst(), "url", url.String())
+	ctxWithDeadline, cancel := context.WithDeadline(ctx, time.Now().Add(waitDeadline))
```

Comment:
Should the deadline be calculated before the `e.ready` recv?

Reply:
I intended to apply it only when using the rate limiter, not when requests are shut down, but I'll change it to apply to both.

Reply:
Done.